### PR TITLE
remove use of `any` type in SqliteStatement

### PR DESCRIPTION
**Pull request by kabentley**
Also, make `IModelDb.requestSnap` and `cancelSnap` @internal

### CODE REVIEW

File: core/backend/src/SqliteStatement.ts [hunk 1]
```diff
@@ -11,21 +11,23 @@ import { ECJsNames, IModelError } from "@bentley/imodeljs-common";
...
- * methods of [SqliteStatement]($backend).
- * @internal
+/** an object returned by [[SqliteStatement.getRow]]
+ * @public
  */
-export interface StringParam {
-  id?: Id64String;
-  guid?: GuidString;
+export interface RowValue {
+  [propName: string]: Uint8Array | number | string;
```
Comment (pmconne) [inline core/backend/src/SqliteStatement.ts:18]: `| undefined` for null?

Comment (kabentley) [reply core/backend/src/SqliteStatement.ts:18]: we don't add it to the result if it's null, apparently.

File: core/backend/src/SqliteStatement.ts [hunk 2]
```diff
@@ -280,15 +280,15 @@ export class SqliteStatement implements IterableIterator<any>, IDisposable {
...
    */
-  public getRow(): any {
+  public getRow(): RowValue {
     const colCount = this.getColumnCount();
-    const row: object = {};
+    const row: RowValue = {};
     const duplicatePropNames = new Map<string, number>();
     for (let i = 0; i < colCount; i++) {
       const sqliteValue = this.getValue(i);
       if (!sqliteValue.isNull) {
```
Comment (kabentley) [inline core/backend/src/SqliteStatement.ts:289]: see here

File: core/backend/src/SqliteStatement.ts (hunk 1 above)
Comment (pmconne) [reply core/backend/src/SqliteStatement.ts:18]: Right but if I do `SELECT a, b FROM c` and get back `row`, if a is null then `row.a` will be undefined. Type should account for that.

Comment (kabentley) [reply core/backend/src/SqliteStatement.ts:18]: I'm always confused about the difference between "not present" and "undefined". Does `typeof row.a` return `"undefined"` if there's no `a` in `row`?

Comment (pmconne) [reply core/backend/src/SqliteStatement.ts:18]: Yes. `undefined` means "no value".

Comment (kabentley) [reply core/backend/src/SqliteStatement.ts]: added

File: core/backend/src/SqliteStatement.ts [hunk 3]
```diff
@@ -11,21 +11,23 @@ import { ECJsNames, IModelError } from "@bentley/imodeljs-common";
...
 
 /** parameter Index (1-based), or name of the parameter (including the initial ':', '@' or '$')
  * @public
  */
 export type BindParameter = number | string;
 
+/** A value that can be bound with [[SqliteStatement.bindValue]] or [[SqliteStatement.bindValues]]
+ * @public
+ */
+export type ParameterValue = undefined | null | number | boolean | string | Uint8Array | { id?: Id64String, guid?: GuidString };
```
Comment (pmconne) [inline core/backend/src/SqliteStatement.ts:29]: Following would more accurately enforce that exactly one of id or guid is permitted.
`| { id: Id64String } | { guid: GuidString }`
Weird to add `null` here where we discourage its use everywhere else and it behaves exactly the same as `undefined`? Or trying to match SQL concept of NULL?

Comment (kabentley) [reply core/backend/src/SqliteStatement.ts:29]: null is accepted in `bindValue`, so I added it. But, probably a good idea to give compile errors so it isn't propagated.

Comment (kabentley) [reply core/backend/src/SqliteStatement.ts]: i removed null. @khanaffan was using it in `NativeAppStorage`, but i don't think that was a good idea.

**Review by pmconne — APPROVED** [review]

**Pull request merged**